### DISCUSSION TITLE
"bitcoin-" prefix in crate names

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1805] [General]
I was just messing around on crates.io and I see that there is a lad who is doing a re-write (transpile) of core into Rust. He has released a whole bunch (> 70) crates on crates.io with the prefix `bitcoin-`.

Going forward we should not assume the prefix implies anything to anyone (as we did while naming `bitcoin-hex`) except that the crate has something to do with bitcoin.

**Comment by Kixunil** [comment]
Heh, proved my point. :) I wonder if folks would be open to reconsider `bitcoin-hex` naming given current situation.

I actually found two people with multiple `bitcoin-` crates: klebs6-s  and joegesualdo

**Reply by apoelstra** [reply]
Well my first alternate suggestion wouldn've been `rust-bitcoin-hex` :P 

I'd still prefer to use bitcoin-hex TBH, unless we actually get scooped on it.

But yes, I care less now about trying to signal stuff via crate names, now that there's another legit bitcoin project out there.

**Comment by stevenroose** [comment]
My opinion would be decouple bitcoin-hex from bitcoin so that it's useful as a generic hex crate and give it a separate repo here in the org. We could give it any name like `epichex` or `hex15` or something.

**Reply by Kixunil** [reply]
That's exactly what I proposed originally (and would still prefer). `hex-conservative` was one of the suggestions.

**Reply by apoelstra** [reply]
I'm cool with `hex-conservative` if that's what we need to move forward :)

**Reply by tcharding** [reply]
Are you good with `hex-conservative` @stevenroose? If yes, could you create a repo please @apoelstra and I'll PR the hex stuff into it from  #1573

**Reply by stevenroose** [reply]
What a politically suggestive name :sweat_smile: But I guess I prefer an suboptimal general name over bitcoin-hex.  :)

I don't recall how it's done exactly, but there's a way to make sure the crate name is just `hex` inside the code so you don't have to do `use hex_conservative::..`, but just `use hex::..`. Something in Cargo.toml does that.

**Reply by apoelstra** [reply]
@tcharding https://github.com/rust-bitcoin/hex-conservative

**Reply by Kixunil** [reply]
@apoelstra I suggest you grab the crates.io name to avoid problems like the last time.

**Reply by apoelstra** [reply]
Yes, I did :).

**Reply by tcharding** [reply]
> I don't recall how it's done exactly, but there's a way to make sure the crate name is just `hex` inside the code so you don't have to do `use hex_conservative::..`, but just `use hex::..`. Something in Cargo.toml does that.

See `bitcoin/Cargo.toml` for an example of one way to do it (putting `extern bitcoin_hashes as hashes;` in `lib.rs` is the other).
```rust
hashes = { package = "bitcoin_hashes", version = "0.12.0", default-features = false }
```

**Comment by stevenroose** [comment]
I was talking about a way to do it in the lib itself. Like how bitcoin_hashes could make itself show up as hashes. I thought `rust-crypto` did this f.e.. Found it: https://github.com/DaGenix/rust-crypto/blob/master/Cargo.toml#L14

And in docs.rs it also shows up like that: https://docs.rs/rust-crypto/latest/crypto/

So use do `use crypto::SomeLessSafeAndLessEfficientImplThanLibsecp;` instead of `use rust_crypto::...`.

**Reply by tcharding** [reply]
~Oh that's cool.~

EDIT: This is functionally exactly the same as we do in `rust-bitcoin` already.

```
[package]
name = "bitcoin"
```

**Reply by tcharding** [reply]
> Like how bitcoin_hashes could make itself show up as hashes

There is a `hashes` crate already unfortunately so this doesn't help for using crates.io dependencies.

**Reply by stevenroose** [reply]
Well, but in rust-bitcoin the package name and crate name are both "bitcoin".. I'm not suggesting any change for hashes.

I'm suggesting if you're creating a `bitcoin-hex` crate, the package name should be `hex` in `Cargo.toml`.

**Reply by Kixunil** [reply]
It can't because `hex` crate already exists on crates.io.

**Reply by stevenroose** [reply]
You sure about that? Cuz there is a crypto crate too (https://crates.io/crates/crypto)..

**Reply by stevenroose** [reply]
What I'm saying is that the crate name and package name can be different. You can have `crates.io/crate/XXX` but have `extern crate YYY` and in docs.rs it will be `docs.rs/XXX/latest/YYY`

**Reply by tcharding** [reply]
I had another go and could not get it to work, perhaps through up a demo PR if you think its definitely possible?

**Reply by Kixunil** [reply]
Oh, I see now, it can be done by setting:

```toml
[lib]
name = "hex"
```

in `Cargo.toml`. However I think it could be very confusing, so downstream users should probably do it themselves. Also I don't know what would happen if someone used both libraries.

**Reply by tcharding** [reply]
Oh, I did exactly that yesterday and it didn't work. Anyways, it feels kinda wrong to me to set our crate to have the same name as someone else's, they got their first, its their name (just opinion).
